File: kahadb/src/main/java/org/apache/kahadb/replication/ReplicationSupport.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.kahadb.replication;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.util.zip.Adler32;
import java.util.zip.Checksum;

import org.apache.kahadb.journal.Location;
import org.apache.kahadb.replication.pb.PBFileInfo;
import org.apache.kahadb.replication.pb.PBJournalLocation;

public class ReplicationSupport {
    
    static public PBJournalLocation convert(Location loc) {
        if( loc==null ) {
            return null;
        }
        return new PBJournalLocation().setFileId(loc.getDataFileId()).setOffset(loc.getOffset());
    }
    
    static public Location convert(PBJournalLocation location) {
        Location rc = new Location();
        rc.setDataFileId(location.getFileId());
        rc.setOffset(location.getOffset());
        return rc;
    }


    static public long copyAndChecksum(File input, File output) throws IOException {
        FileInputStream is = null;
        FileOutputStream os = null;
        try {
            is = new FileInputStream(input);
            os = new FileOutputStream(output);

            byte buffer[] = new byte[1024 * 4];
            int c;

            Checksum checksum = new Adler32();
            while ((c = is.read(buffer)) >= 0) {
                os.write(buffer, 0, c);
                checksum.update(buffer, 0, c);
            }
            return checksum.getValue();

        } finally {
            try {
                is.close();
            } catch(Throwable e) {
            }
            try {
                os.close();
            } catch(Throwable e) {
            }
        }
    }

    public static PBFileInfo createInfo(String name, File file, long start, long length) throws IOException {
        PBFileInfo rc = new PBFileInfo();
        rc.setName(name);
        rc.setChecksum(checksum(file, start, length));
        rc.setStart(start);
        rc.setEnd(length);
        return rc;
    }

    public static long checksum(File file, long start, long end) throws IOException {
        RandomAccessFile raf = new RandomAccessFile(file, "r");
        try {
            Checksum checksum = new Adler32();
            byte buffer[] = new byte[1024 * 4];
            int c;
            long pos = start;
            raf.seek(start);

            while (pos < end && (c = raf.read(buffer, 0, (int) Math.min(end - pos, buffer.length))) >= 0) {
                checksum.update(buffer, 0, c);
                pos += c;
            }

            return checksum.getValue();
        } finally {
            try {
                raf.close();
            } catch (Throwable e) {
            }
        }
    }

}

File: activemq-leveldb-store/src/main/scala/org/apache/activemq/leveldb/replicated/ReplicationSupport.scala
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.activemq.leveldb.replicated

import org.fusesource.hawtbuf.Buffer._
import java.util.concurrent._
import java.nio.MappedByteBuffer
import sun.nio.ch.DirectBuffer
import java.io.{RandomAccessFile, File}
import java.nio.channels.FileChannel
import java.util.concurrent.atomic.AtomicInteger
import org.fusesource.hawtdispatch._
import org.apache.activemq.leveldb.util.FileSupport._
import org.apache.activemq.leveldb.LevelDBClient
import scala.collection.immutable.TreeMap

object ReplicationSupport {

  val WAL_ACTION = ascii("wal")
  val LOGIN_ACTION= ascii("LevelDB Store Replication v1:login")
  val SYNC_ACTION = ascii("sync")
  val GET_ACTION = ascii("get")
  val ACK_ACTION = ascii("ack")
  val OK_ACTION = ascii("ok")
  val DISCONNECT_ACTION = ascii("disconnect")
  val ERROR_ACTION = ascii("error")
  val LOG_DELETE_ACTION = ascii("rm")

  def unmap(buffer:MappedByteBuffer ) {
    try {
      buffer.asInstanceOf[DirectBuffer].cleaner().clean();
    } catch {
      case ignore:Throwable =>
    }
  }

  def map(file:File, offset:Long, length:Long, readOnly:Boolean) = {
    val raf = new RandomAccessFile(file, if(readOnly) "r" else "rw");
    try {
      val mode = if (readOnly) FileChannel.MapMode.READ_ONLY else FileChannel.MapMode.READ_WRITE
      raf.getChannel().map(mode, offset, length);
    } finally {
      raf.close();
    }
  }

  def stash(directory:File) {
    directory.mkdirs()
    val tmp_stash = directory / "stash.tmp"
    val stash = directory / "stash"
    stash.recursiveDelete
    tmp_stash.recursiveDelete
    tmp_stash.mkdirs()
    copy_store_dir(directory, tmp_stash)
    tmp_stash.renameTo(stash)
  }

  def copy_store_dir(from:File, to:File) = {
    val log_files = LevelDBClient.find_sequence_files(from, LevelDBClient.LOG_SUFFIX)
    if( !log_files.isEmpty ) {
      val append_file = log_files.last._2
      for( file <- log_files.values ; if file != append_file) {
        file.linkTo(to / file.getName)
        val crc_file = file.getParentFile / (file.getName+".crc32" )
        if( crc_file.exists() ) {
          crc_file.linkTo(to / crc_file.getName)
        }
      }
      append_file.copyTo(to / append_file.getName)
    }

    val index_dirs = LevelDBClient.find_sequence_files(from, LevelDBClient.INDEX_SUFFIX)
    if( !index_dirs.isEmpty ) {
      val index_file = index_dirs.last._2
      var target = to / index_file.getName
      target.mkdirs()
      LevelDBClient.copyIndex(index_file, target)
    }
  }

  def stash_clear(directory:File) {
    val stash = directory / "stash"
    stash.recursiveDelete
  }

  def unstash(directory:File) {
    val tmp_stash = directory / "stash.tmp"
    tmp_stash.recursiveDelete
    val stash = directory / "stash"
    if( stash.exists() ) {
      delete_store(directory)
      copy_store_dir(stash, directory)
      stash.recursiveDelete
    }
  }

  def delete_store(directory: File) {
    // Delete any existing files to make space for the stash we will be restoring..
    var t: TreeMap[Long, File] = LevelDBClient.find_sequence_files(directory, LevelDBClient.LOG_SUFFIX)
    for (entry <- t) {
      val file = entry._2
      file.delete()
      val crc_file = directory / (file.getName+".crc32" )
      if( crc_file.exists() ) {
        crc_file.delete()
      }
    }
    for (file <- LevelDBClient.find_sequence_files(directory, LevelDBClient.INDEX_SUFFIX)) {
      file._2.recursiveDelete
    }
  }
}
